---
title: "Crash course introduction to prediction computation"
date: "`r format(Sys.time(), '%B %Y')`"
author: "Paul Yousefi, PhD, MPH"
output:
  html_document:
    css: style.css
    highlight: tango
  pdf_document: default
---

# Before we start

This lab will performed in R and will use the following packages:

__Add downloads__


We'll be using data from Tibshirani et al. that is publicly available on the gene expression ombibus (GEO) website

All course material, including the data and code used for this lab practical, is available for you to download here:

__url for where to download data__


# Goals 

* Partitioning data into training and testing sets
* Evaluating performance of risk scores
* Fitting models in training data
* Predicting outputs from those models in the testing data
* Quantifying model prediction performance


```{r, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(
comment = ">",
error = FALSE,
tidy = FALSE,
echo = TRUE, 
warning = FALSE, 
message = FALSE, 
cache=F)
```
```{r call_source, echo = F}
path <- "."
read_chunk(file.path(path, "prediction-practical-2019-ser-sourcecode.r"))
```
```{r globals, echo = F, results="hide", message = F, warning = F}
```


# Getting started 

To start, I'll load our data into active memory and have a look at what's available:

```{r load, echo = T} 
```
```{r ls, eval = F, echo=T} 
```
> [1] "meth"    "samples"


So we have two data objects:

* `meth` with DNA methylation data

* `samples` with other phenotype information on the participants of this study


Let's get a better sense of the variables available in `samples`:
 
```{r qc1, eval = T, echo=T} 
```

The `smoking` variable has 3 categories, but it's easiest to begin with a binary outcome so let's focus on the `ever.smoke` variable that collapses the __current__ and __former__ subjects into a single category

* When I talk about predicting smoking going from now on I'll be referring to this `ever.smoke`variable 

# Applying risk scores

### Single variable scores

The simplest type of risk score we can use for prediction is just a single individual variable. The site `cg05575921` in the _AHRR_ gene has consistently been the CpG with methylation showing the strongest association with smoking in several studies looking broadly across the genome.

Perhaps the methylation levels of this site would be sufficient to predict whether someone has been a smoker. To see, let's begin by adding this CpG site as a variable to our phenotype data object `samples`:

```{r ahrr} 
```

We can use a package called `pROC` to see how well different values of our `ahrr` variable explain smoking status:

```{r roc1} 
```

We can also visualize our results by using the `pROC` package's `plot.roc()` function on the saved output

```{r plot.roc}
```

### Weighted risk scores from published coefficients

Another common approach in omic prediction is to apply a risk score using information from multiple loci or omic measures weighted by previously reported magnitudes of association observed between those features and the outcome of interest. This has perhaps been most often performed using genetic data to compose 'polygenic risk scores', but can easily be extended to other types of data input.

For example, in the context of DNA methylation data we can define and subsequently apply a smoking score derived from the published coefficients of the largest blood epigenome wide association study (EWAS) meta-analysis to date in Joehanes et al. 2016:

>Joehanes R, Just AC, Marioni RE, Pilling LC, Reynolds LM, Mandaviya PR et al. Epigenetic Signatures of Cigarette Smoking. Circ Cardiovasc Genet 2016;

* The coefficients from their models were distributed in their Supplemental Table 2 that I previously saved and load into R below: 

```{r load.joehanes} 
```

The `joehanes` object has summary information on the `r nrow(joehanes)` CpGs that were significant at a Bonferroni p-value threshold in the original meta-analysis and that were available in our methylation dataset

```{r joehanes_str} 
```

Let's restrict our big methylation data object, `meth`, to just the CpGs that are in the `joehanes` list. This keeps the CpGs that we expect to be most related to smoking behavior, in the correct format, while reducing the size of the data were working with:

```{r subset.meth} 
```

To compute an individual's risk score using these coefficients, we need to take the sum of each coefficient multiplied by the participant's value at the corresponding variable, for example:

$$\hat{Y}_{Joehanes Score} = \sum_{i}^{2617 CpGs} \hat{\beta}_{i} X_{i}$$

* Where $\hat{\beta}_{i}$ are the individual previously estimated Joehanes et al. coefficients and $X_i$ are their corresponding variables (CpG site measurements in this case).

Equivalently, we can compute the exact same quantity, $\hat{Y}_{Joehanes Score}$, more simply using matrix multiplication:

$$\hat{Y}_{Joehanes Score} = X \hat{\beta}$$

* Where X is an $N x P$ matrix of all P variables being used and $\hat{\beta}$ is a corresponding vector of all Joehanes et al. coefficients.


To implement this, lets start by making a named vector of the `joehanes` coefficients:

```{r make_coefs} 
```

We can then use matrix multiplication against our observed DNA methylation values to get our $\hat{Y}_{Joehanes Score}$ values:

```{r apply_coefs} 
```

By adding this output as a variable to our `samples` data, we can again use the `pROC` package to evaluate and visualize the prediction performance of this score:


```{r add_yhat}
```

```{r plot.roc.again}
```

Does the joehanes score predict never/ever smoking better than just cg05575921 alone?

```{r comp_roc}
```

No, it doesn't appear to!

$\vspace*{2\baselineskip}$

# Training a novel predictor


```{r data.partition}
```


```{r data.subset}
```

