---
title: "Crash course introduction to prediction computation"
date: "`r format(Sys.time(), '%B %Y')`"
author: "Paul Yousefi, PhD, MPH"
output:
  html_document:
    css: style.css
    highlight: tango
  pdf_document: default
---

# Before we start

This lab will performed in R and will use the following packages:

__Add downloads__


We'll be using data from Tibshirani et al. that is publicly available on the gene expression ombibus (GEO) website

All course material, including the data and code used for this lab practical, is available for you to download here:

__url for where to download data__


# Goals 

* Partitioning data into training and testing sets
* Evaluating performance of risk scores
* Fitting models in training data
* Predicting outputs from those models in the testing data
* Quantifying model prediction performance


```{r, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(
comment = ">",
error = FALSE,
tidy = FALSE,
echo = TRUE, 
warning = FALSE, 
message = FALSE, 
cache=F)
```
```{r call_source, echo = F}
path <- "."
read_chunk(file.path(path, "prediction-practical-2019-ser-sourcecode.r"))
```
```{r globals, echo = F, results="hide", message = F, warning = F}
```


# Getting started 

To start, I'll load our data into active memory and have a look at what's available:

```{r load, echo = T} 
```
```{r ls, eval = F, echo=T} 
```
> [1] "meth"    "samples"


So we have two data objects:

* `meth` with DNA methylation data

* `samples` with other phenotype information on the participants of this study


Let's get a better sense of the variables available in `samples`:
 
```{r qc1, eval = T, echo=T} 
```

The `smoking` variable has 3 categories, but it's easiest to begin with a binary outcome so let's focus on the `ever.smoke` variable that collapses the __current__ and __former__ subjects into a single category

* When I talk about predicting smoking going from now on I'll be referring to this `ever.smoke`variable 

# Applying risk scores

### Single variable scores

The simplest type of risk score we can use for prediction is just a single individual variable. The site `cg05575921` in the _AHRR_ gene has consistently been the CpG with methylation showing the strongest association with smoking in several studies looking broadly across the genome.

Perhaps the methylation levels of this site would be sufficient to predict whether someone has been a smoker. To see, let's begin by adding this CpG site as a variable to our phenotype data object `samples`:

```{r ahrr} 
```

We can use a package called `pROC` to see how well different values of our `ahrr` variable explain smoking status:

```{r roc1} 
```

We can also visualize our results by using the `pROC` package's `plot.roc()` function on the saved output

```{r plot.roc}
```

### Weighted risk scores from published coefficients





